File: apps/app-frontend/src/helpers/noctrinth-crash.ts
/**
 * The launcher's own reading of a crash.
 *
 * Upstream sends the live console to mclo.gs and shows what it makes of it,
 * which is the better answer for anything unusual — and no answer at all
 * without a connection, or for the crash report and the JVM's own error file,
 * which is where the cause is usually written in plain sight.
 *
 * This is the local half: rules that run against those files as well, on this
 * machine, and fixes that are pages of this launcher rather than advice. See
 * `packages/app-lib/src/api/crash_analysis.rs` for why the words a player reads
 * are not in the Rust.
 */
import { invoke } from '@tauri-apps/api/core'

export type CrashSeverity = 'note' | 'warning' | 'critical'
export type CrashSourceKind = 'crash_report' | 'log' | 'jvm_error'

export interface CrashFinding {
	rule: string
	severity: CrashSeverity
	source: CrashSourceKind
	source_name: string
	evidence: string
	values: Record<string, string>
}

export interface CrashSourceFile {
	kind: CrashSourceKind
	name: string
	modified: number
}

export interface CrashDiagnosis {
	findings: CrashFinding[]
	sources: CrashSourceFile[]
}

const EMPTY: CrashDiagnosis = { findings: [], sources: [] }

/** Reads the files the last run left behind and reports what it recognised. */
export async function analyzeInstanceCrash(instanceId: string): Promise<CrashDiagnosis> {
	try {
		return await invoke<CrashDiagnosis>('plugin:noctrinth-crash|crash_analyze_instance', {
			instanceId,
		})
	} catch (error) {
		// A diagnosis that cannot be made is not worth a notification: the log
		// the player came here to read is on screen either way.
		console.warn('Could not read the crash files for a diagnosis', error)
		return EMPTY
	}
}

/** The same rules, against text already in hand. */
export async function analyzeCrashText(
	text: string,
	kind: CrashSourceKind,
	sourceName: string,
): Promise<CrashDiagnosis> {
	if (!text.trim()) return EMPTY

	try {
		return await invoke<CrashDiagnosis>('plugin:noctrinth-crash|crash_analyze_text', {
			text,
			kind,
			sourceName,
		})
	} catch (error) {
		console.warn('Could not analyse the log', error)
		return EMPTY
	}
}

/** Which of two diagnoses to lead with, worst first. */
const SEVERITY_ORDER: Record<CrashSeverity, number> = {
	critical: 0,
	warning: 1,
	note: 2,
}

export function sortFindings(findings: CrashFinding[]): CrashFinding[] {
	return [...findings].sort((a, b) => SEVERITY_ORDER[a.severity] - SEVERITY_ORDER[b.severity])
}

File: apps/app/src/api/noctrinth_crash.rs
//! The launcher's own reading of what a bad run left behind.
//!
//! A plugin of the fork's own so that upstream's stay upstream's. What it
//! exposes is the analysis in `theseus::crash_analysis` — see that module for
//! why the words a player reads are not in it.

use crate::api::Result;
use theseus::crash_analysis::{self, CrashDiagnosis, CrashSourceKind};

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("noctrinth-crash")
        .invoke_handler(tauri::generate_handler![
            crash_analyze_instance,
            crash_analyze_text,
        ])
        .build()
}

/// Reads the crash report, the JVM's own error file and the tail of the log an
/// instance last wrote, and reports what it recognised.
#[tauri::command]
pub async fn crash_analyze_instance(
    instance_id: &str,
) -> Result<CrashDiagnosis> {
    Ok(crash_analysis::analyze_instance(instance_id).await?)
}

/// The same rules, against text the interface already has in hand: the log
/// being looked at, or the console of a run whose files are not written yet.
#[tauri::command]
pub async fn crash_analyze_text(
    text: String,
    kind: CrashSourceKind,
    source_name: String,
) -> Result<CrashDiagnosis> {
    Ok(crash_analysis::analyze_text(&text, kind, &source_name))
}
